File: backend/app/services/document_processor.py
"""Service de découpage (chunking) de documents texte."""
from __future__ import annotations

import re


def chunk_text(
    text: str,
    chunk_size: int = 1500,
    overlap: int = 150,
) -> list[tuple[str, int]]:
    """Découpe un texte en chunks avec chevauchement.

    Returns:
        Liste de tuples (contenu_chunk, nb_tokens_approximatif)
    """
    # Nettoyage basique
    text = text.strip()
    if not text:
        return []

    # Séparateurs par ordre de priorité
    separators = ["\n\n\n", "\n\n", "\n", ". ", " "]
    chunks: list[tuple[str, int]] = []

    paragraphs = _split_by_separators(text, separators, chunk_size)

    buffer = ""
    for para in paragraphs:
        if not para.strip():
            continue
        candidate = (buffer + "\n\n" + para).strip() if buffer else para.strip()
        tokens = _estimate_tokens(candidate)

        if tokens <= chunk_size:
            buffer = candidate
        else:
            if buffer:
                chunks.append((buffer, _estimate_tokens(buffer)))
            # Si le paragraphe seul dépasse chunk_size, le forcer
            if _estimate_tokens(para) > chunk_size:
                sub_chunks = _force_split(para, chunk_size)
                chunks.extend(sub_chunks)
                buffer = ""
            else:
                buffer = para.strip()

    if buffer:
        chunks.append((buffer, _estimate_tokens(buffer)))

    # Applique le chevauchement
    if overlap > 0 and len(chunks) > 1:
        chunks = _apply_overlap(chunks, overlap)

    return chunks


def _split_by_separators(text: str, separators: list[str], max_tokens: int) -> list[str]:
    """Tente de couper aux séparateurs naturels."""
    for sep in separators:
        parts = text.split(sep)
        if all(_estimate_tokens(p) <= max_tokens for p in parts):
            return parts
    return [text]


def _force_split(text: str, chunk_size: int) -> list[tuple[str, int]]:
    """Coupe brutalement un texte trop long par mots."""
    words = text.split()
    chunks = []
    buffer_words: list[str] = []
    current_tokens = 0

    for word in words:
        wt = max(1, len(word) // 4)
        if current_tokens + wt > chunk_size and buffer_words:
            chunk = " ".join(buffer_words)
            chunks.append((chunk, _estimate_tokens(chunk)))
            buffer_words = []
            current_tokens = 0
        buffer_words.append(word)
        current_tokens += wt

    if buffer_words:
        chunk = " ".join(buffer_words)
        chunks.append((chunk, _estimate_tokens(chunk)))

    return chunks


def _apply_overlap(
    chunks: list[tuple[str, int]],
    overlap: int,
) -> list[tuple[str, int]]:
    """Ajoute un chevauchement entre chunks consécutifs."""
    result: list[tuple[str, int]] = [chunks[0]]
    for i in range(1, len(chunks)):
        prev_text = chunks[i - 1][0]
        curr_text = chunks[i][0]
        # Prend les derniers mots du chunk précédent comme préfixe
        prev_words = prev_text.split()
        overlap_words = prev_words[-overlap:] if len(prev_words) > overlap else prev_words
        prefix = " ".join(overlap_words)
        merged = (prefix + "\n" + curr_text).strip()
        result.append((merged, _estimate_tokens(merged)))
    return result


def _estimate_tokens(text: str) -> int:
    """Estimation rapide du nombre de tokens (≈ 4 caractères par token)."""
    return max(1, len(text) // 4)


def process_file(path: str, content: str) -> list[tuple[str, int]]:
    """Point d'entrée principal pour le traitement d'un fichier."""
    # Détecte si c'est du code source → meilleur découpage par blocs
    code_extensions = {".py", ".rs", ".js", ".ts", ".go", ".java", ".c", ".cpp", ".sh"}
    import os
    ext = os.path.splitext(path)[1].lower()

    if ext in code_extensions:
        return _chunk_code(content)

    return chunk_text(content)


def _chunk_code(content: str, chunk_size: int = 1500) -> list[tuple[str, int]]:
    """Découpage intelligent pour du code source (par fonctions/classes)."""
    # Sépare sur les blocs de fonctions/classes (lignes vides + définition)
    blocks = re.split(r"\n(?=\s*(?:def |class |fn |function |func |async def |pub fn |impl ))", content)
    chunks: list[tuple[str, int]] = []
    buffer = ""

    for block in blocks:
        candidate = (buffer + "\n\n" + block).strip() if buffer else block.strip()
        tokens = _estimate_tokens(candidate)

        if tokens <= chunk_size:
            buffer = candidate
        else:
            if buffer:
                chunks.append((buffer, _estimate_tokens(buffer)))
            if _estimate_tokens(block) > chunk_size:
                chunks.extend(_force_split(block, chunk_size))
                buffer = ""
            else:
                buffer = block.strip()

    if buffer:
        chunks.append((buffer, _estimate_tokens(buffer)))

    return chunks if chunks else chunk_text(content)

File: backend-rs/src/services/document_processor.rs
//! Document chunking service — code-aware and paragraph-based text splitting.

use regex::Regex;

/// Estimate token count (~4 chars per token).
fn estimate_tokens(text: &str) -> usize {
    (text.len() / 4).max(1)
}

/// Split text into chunks with overlap. Returns Vec<(content, token_count)>.
pub fn chunk_text(text: &str, chunk_size: usize, overlap: usize) -> Vec<(String, usize)> {
    let text = text.trim();
    if text.is_empty() {
        return Vec::new();
    }

    let separators = ["\n\n\n", "\n\n", "\n", ". ", " "];
    let paragraphs = split_by_separators(text, &separators, chunk_size);

    let mut chunks: Vec<(String, usize)> = Vec::new();
    let mut buffer = String::new();

    for para in &paragraphs {
        let para = para.trim();
        if para.is_empty() {
            continue;
        }
        let candidate = if buffer.is_empty() {
            para.to_string()
        } else {
            format!("{}\n\n{}", buffer, para)
        };
        let tokens = estimate_tokens(&candidate);

        if tokens <= chunk_size {
            buffer = candidate;
        } else {
            if !buffer.is_empty() {
                let t = estimate_tokens(&buffer);
                chunks.push((buffer.clone(), t));
            }
            if estimate_tokens(para) > chunk_size {
                chunks.extend(force_split(para, chunk_size));
                buffer.clear();
            } else {
                buffer = para.to_string();
            }
        }
    }

    if !buffer.is_empty() {
        let t = estimate_tokens(&buffer);
        chunks.push((buffer, t));
    }

    // Apply overlap
    if overlap > 0 && chunks.len() > 1 {
        chunks = apply_overlap(chunks, overlap);
    }

    chunks
}

/// Code-aware chunking: splits on function/class definitions.
pub fn chunk_code(content: &str, chunk_size: usize) -> Vec<(String, usize)> {
    // Rust regex doesn't support lookahead — split on lines starting with definitions
    let re = Regex::new(r"\n\s*(?:def |class |fn |function |func |async def |pub fn |impl )").unwrap();

    // Manual split preserving the delimiter at the start of each block
    let mut blocks: Vec<String> = Vec::new();
    let mut last = 0;
    for m in re.find_iter(content) {
        if m.start() > last {
            blocks.push(content[last..m.start()].to_string());
        }
        last = m.start() + 1; // keep \n with the next block
    }
    if last < content.len() {
        blocks.push(content[last..].to_string());
    }
    if blocks.is_empty() {
        blocks.push(content.to_string());
    }

    let mut chunks: Vec<(String, usize)> = Vec::new();
    let mut buffer = String::new();

    for block in blocks {
        let candidate = if buffer.is_empty() {
            block.trim().to_string()
        } else {
            format!("{}\n\n{}", buffer, block.trim())
        };
        let tokens = estimate_tokens(&candidate);

        if tokens <= chunk_size {
            buffer = candidate;
        } else {
            if !buffer.is_empty() {
                let t = estimate_tokens(&buffer);
                chunks.push((buffer.clone(), t));
            }
            if estimate_tokens(&block) > chunk_size {
                chunks.extend(force_split(&block, chunk_size));
                buffer.clear();
            } else {
                buffer = block.trim().to_string();
            }
        }
    }

    if !buffer.is_empty() {
        let t = estimate_tokens(&buffer);
        chunks.push((buffer, t));
    }

    if chunks.is_empty() {
        chunk_text(content, chunk_size, 0)
    } else {
        chunks
    }
}

/// Entry point: detect file type and dispatch to appropriate chunker.
pub fn process_file(path: &str, content: &str, chunk_size: usize, overlap: usize) -> Vec<(String, usize)> {
    let code_extensions = [".py", ".rs", ".js", ".ts", ".go", ".java", ".c", ".cpp", ".sh"];
    let ext = path.rfind('.').map(|i| &path[i..]).unwrap_or("");
    let ext_lower = ext.to_lowercase();

    if code_extensions.iter().any(|e| ext_lower == *e) {
        let mut chunks = chunk_code(content, chunk_size);
        if overlap > 0 && chunks.len() > 1 {
            chunks = apply_overlap(chunks, overlap);
        }
        chunks
    } else {
        chunk_text(content, chunk_size, overlap)
    }
}

// ── Internal helpers ──

fn split_by_separators(text: &str, separators: &[&str], max_tokens: usize) -> Vec<String> {
    for sep in separators {
        let parts: Vec<&str> = text.split(sep).collect();
        if parts.iter().all(|p| estimate_tokens(p) <= max_tokens) {
            return parts.into_iter().map(String::from).collect();
        }
    }
    vec![text.to_string()]
}

fn force_split(text: &str, chunk_size: usize) -> Vec<(String, usize)> {
    let words: Vec<&str> = text.split_whitespace().collect();
    let mut chunks = Vec::new();
    let mut buf_words: Vec<&str> = Vec::new();
    let mut current_tokens = 0usize;

    for word in words {
        let wt = (word.len() / 4).max(1);
        if current_tokens + wt > chunk_size && !buf_words.is_empty() {
            let chunk = buf_words.join(" ");
            let t = estimate_tokens(&chunk);
            chunks.push((chunk, t));
            buf_words.clear();
            current_tokens = 0;
        }
        buf_words.push(word);
        current_tokens += wt;
    }

    if !buf_words.is_empty() {
        let chunk = buf_words.join(" ");
        let t = estimate_tokens(&chunk);
        chunks.push((chunk, t));
    }

    chunks
}

fn apply_overlap(chunks: Vec<(String, usize)>, overlap: usize) -> Vec<(String, usize)> {
    let mut result: Vec<(String, usize)> = vec![chunks[0].clone()];

    for i in 1..chunks.len() {
        let prev_text = &chunks[i - 1].0;
        let curr_text = &chunks[i].0;
        let prev_words: Vec<&str> = prev_text.split_whitespace().collect();
        let start = if prev_words.len() > overlap {
            prev_words.len() - overlap
        } else {
            0
        };
        let overlap_words = &prev_words[start..];
        let prefix = overlap_words.join(" ");
        let merged = format!("{}\n{}", prefix, curr_text).trim().to_string();
        let t = estimate_tokens(&merged);
        result.push((merged, t));
    }

    result
}
